src/v/raft/mux_state_machine.h

```diff
@@ -281,17 +281,22 @@ ss::future<> mux_state_machine<T...>::apply(model::record_batch b) {
           *state);
 
         return result_f.then([this, last_offset](std::error_code ec) {
-            auto f = _mutex.with([this, last_offset, ec] {
-                if (auto it = _promises.find(last_offset);
-                    it != _promises.end()) {
-                    it->second.set_value(ec);
-                }
-            });
-            if (!_persist_last_applied) {
-                return f;
-            }
-            return f.then(
-              [this, last_offset] { return write_last_applied(last_offset); });
+            return _mutex
+              .with([this, last_offset, ec] {
+                  if (auto it = _promises.find(last_offset);
+                      it != _promises.end()) {
+                      it->second.set_value(ec);
+                  }
```

Comment:
I'm sure this was asked at some point in the past, but why is a mutex being used to protect an already non-interruptible code block?

Reply:
this is related with the fact that offset is known after replicate finishes, mutex is used to make replicate and its continuation atomic.